### PR TITLE
Add user agent and scopes

### PR DESCRIPTION
**Pull request by tomhjp**
# Overview

Adds user agent to MongoDB Atlas requests for tracking usage, and adds the ability to specify scopes for created users. Both requested by MongoDB.

# Design of Change

Adding scopes required a client upgrade, which in turn required some updates to calling `DeleteUser`. This means we can support managing users on non-"admin" database names now, as per https://docs.atlas.mongodb.com/reference/api/database-users-create-a-user/#request-body-parameters. The Vault docs state that Atlas does not support non-"admin" database names, but the Atlas docs seem to suggest it does, so perhaps it has been updated since.

Alongside `databaseName` being supported in delete statements, this PR also requires a docs update for adding `scopes` to creation statements, detailed in the same docs link as above. I haven't written a docs PR yet because I wanted to get some feedback on this PR first.

On setting user agent, unfortunately I don't think we can get at a `PluginEnvironment` struct, so I'm not sure we can use the slightly more detailed `useragent.PluginString`. though I would like a second opinion on that.

Finally, I'm not sure whether I can really test the user agent update, because communication is handled internally to the MongoDB client and uses TLS. Any thoughts on that appreciated.

# Related Issues/Pull Requests
https://github.com/hashicorp/vault-plugin-secrets-mongodbatlas/pull/13

# Contributor Checklist
- [ ] Add relevant docs to upstream Vault repository, or sufficient reasoning why docs won’t be added yet
- [x] Add output for any tests not ran in CI to the PR description (eg, acceptance tests)
- [x] Backwards compatible

Acceptance tests passing:

```bash
go test ./...
ok      github.com/hashicorp/vault-plugin-database-mongodbatlas 176.564s
```

### CODE REVIEW

File: connection_producer.go [hunk 1]
```diff
@@ -2,11 +2,12 @@ package mongodbatlas
 
 import (
 	"context"
+	"github.com/hashicorp/vault/sdk/helper/useragent"
 	"sync"
 
 	"github.com/Sectorbob/mlab-ns2/gae/ns/digest"
 	"github.com/hashicorp/vault/sdk/database/helper/connutil"
-	"github.com/mongodb/go-client-mongodb-atlas/mongodbatlas"
+	"go.mongodb.org/atlas/mongodbatlas"
```
Comment (jasonodonnell) [inline connection_producer.go:10]: Was this change intended to support scopes? Can our client be extended if so?

Comment (tomhjp) [reply connection_producer.go:10]: Ah I should have explained in the description, this is actually the same module, but they have updated the package path since the version we were on before: https://github.com/mongodb/go-client-mongodb-atlas/pull/87

File: connection_producer.go [hunk 2]
```diff
@@ -59,6 +60,7 @@ func (c *mongoDBAtlasConnectionProducer) Connection(_ context.Context) (interfac
 	if err != nil {
 		return nil, err
 	}
+	client.UserAgent = useragent.String()
```
[the remark this answers is not in the document]
Comment (kalafut) [reply connection_producer.go:63]: For the common case of the plugin being bundled, the sdk will be the same files that Vault is using which should be (better be!) accurate.

File: mongodbatlas.go [hunk 3]
```diff
@@ -180,7 +182,20 @@ func (m *MongoDBAtlas) DeleteUser(ctx context.Context, req newdbplugin.DeleteUse
 		return newdbplugin.DeleteUserResponse{}, err
 	}
 
-	_, err = client.DatabaseUsers.Delete(ctx, m.ProjectID, req.Username)
+	var databaseUser mongoDBAtlasStatement
+	if len(req.Statements.Commands) > 0 {
+		err = json.Unmarshal([]byte(req.Statements.Commands[0]), &databaseUser)
+		if err != nil {
+			return newdbplugin.DeleteUserResponse{}, fmt.Errorf("error unmarshalling statement %s", err)
```
Comment (kalafut) [inline mongodbatlas.go:189]: Prefer %w for these now?

File: connection_producer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomhjp) [reply connection_producer.go:63]: Within the limitations of a database plugin, AFAIK the only thing we could do to improve this is also include the version of the plugin itself (and add the necessary code/linker flags so the plugin knows its own version).

On a similar note, the `useragent` package doesn't let us include the plugin name unless we have a `*logical.PluginEnvironment` available, so allowing version as a string would be another possible update.

Are either of those worthwhile updates, or do you think we should land the PR as-is?

**Comment by calvn** [comment]
@tomhjp Looks like there are a few conflicts since then, can you resolve those?

**Comment by tomhjp** [comment]
@calvn thanks, I've resolved those conflicts now, and rerun the acceptance tests:

```bash
go test -v ./...
=== RUN   TestIntegrationDatabaseUser_Initialize
--- PASS: TestIntegrationDatabaseUser_Initialize (0.00s)
=== RUN   TestAcceptanceDatabaseUser_CreateUser
    mongodbatlas_test.go:190: Asserting username: v-test-Yl6skiNNN5lRO
--- PASS: TestAcceptanceDatabaseUser_CreateUser (27.67s)
=== RUN   TestAcceptanceDatabaseUser_CreateUserWithSpecialChar
    mongodbatlas_test.go:234: Asserting username: v-test-NaA7Mrpp5zEgJ
--- PASS: TestAcceptanceDatabaseUser_CreateUserWithSpecialChar (31.56s)
=== RUN   TestAcceptanceDatabaseUser_DeleteUser
    mongodbatlas_test.go:283: Asserting username: v-test-G84jxpv7QLfxJ
--- PASS: TestAcceptanceDatabaseUser_DeleteUser (61.46s)
=== RUN   TestAcceptanceDatabaseUser_UpdateUser_Password
    mongodbatlas_test.go:327: Asserting username: testmongouser
    mongodbatlas_test.go:340: Asserting username: testmongouser
--- PASS: TestAcceptanceDatabaseUser_UpdateUser_Password (93.23s)
PASS
ok      github.com/hashicorp/vault-plugin-database-mongodbatlas 215.013s
```

**Review by jasonodonnell — APPROVED** [review]
LGTM
